### ISSUE TITLE
OQS sub projects: Which ones to drop for good

### ISSUE THREAD

**Issue by baentsch** [open-quantum-safe/tsc#2]
This is to initiate a discussion on OQS sub projects: There are many that do not receive any attention in terms of issue-creation, -resolution or other activity. We have declared some as "inactive" to no great(ly visible) disagreement.

Proposal: 
- Designate 3 sub project tiers: Fully supported (at least 2 fully committed maintainers), best effort (maybe a part-time maintainer), inactive (none of us cares).
- Assign all projects to these tiers
- Only discuss/track fully supported projects going forward, e.g., in https://www.openquantumsafe.org/dashboard.html
- Integrate (cross-project) CI only for fully supported projects, incl. support of the same platforms & binaries/distro-packages
- Get input from PQCA members as to which projects they want (& fund)

Strawman proposal: Only `liboqs`, `oqsprovider` and possibly `liboqs-python` receive "fully supported" status.

Input/feedback/alternative suggestions welcome.

**Comment by christianpaquin** [comment]
I agree with the proposed 3 tiers, I think all 3 should get a different "message" (i.e., don't mark "best effort" and "inactive" with the same message, and we want to run some "best effort" cross-CI testing to "best effort" projects (could be prioritized based on available resources). Agreed that we need to urgently reflect the expected level of support in the project's documentation.

**Comment by dstebila** [comment]
Here's the PQCA's working document for their project lifecycle.

**Comment by baentsch** [comment]
> Here's the PQCA's working document for their project lifecycle: https://docs.google.com/document/d/1NV-0vNgXWdc81oqT0jv0C-9Funb8dySS06u90ghF-X4/edit

What concrete impact does this document have on this issue? Is there any timeline for this document to be completed (seems inactive)? Should completion of this document stop this issue from proceeding?

My concrete proposal:
- Suggest to @vsoftco to create GOVERNANCE.md files modeled after those of oqs-provider & liboqs for those projects he's maintaining.
- Drop all sub projects without GOVERNANCE.md files (and hence maintainers) from the list of actively maintained OQS sub projects, e.g., on the Web site (as already done for ssh&openssl-111) as well as [CI, incl. dashboard](https://openquantumsafe.org/dashboard.html).

**Comment by baentsch** [comment]
Thanks for taking a stab at this summary, @dstebila.

I completely agree with columns 1-3 with one request for change: The term "Fully supported" should be changed to "Fully github supported" for `oqs-provider` in the third column as the former may imply support on all channels PQCA has chosen to create (email lists, DM tooling, ...) but that I cannot follow and hence, cannot promise to provide support on. Tagging @thb-sb for commenting on this: If you'd follow and interact with the PQCA comms channels (and keep acting as `oqs-provider` maintainer) (?), it may be OK to leave the entry as-is.

Regarding the forth column, I have to absolutely and adamantly protest the labelling of `oqs-provider` as (LF/PQCA) "Impact Stage": That project is NOT receiving financial nor any other support from PQCA (rather the opposite if I compare how OQS operates before and after the LF take-over: Contributions and procedures have surely become more difficult and cumbersome). Most notably, `oqs-provider` (well, I guess myself as the 95% contributor/maintainer of it) is accordingly not willing to "to cross promote the foundation along with their activities" as per the demand placed on such project according to the [PQCA project classification](https://docs.google.com/document/d/1NV-0vNgXWdc81oqT0jv0C-9Funb8dySS06u90ghF-X4/edit) --> please change over to "Incubation stage" if you need to apply this classification -- that I personally consider pretty much completely unsuitable to OQS:

To justify the latter statement, see these two pretty much contradictory statements:

>  Document that it is being used successfully in production by at least two independent end users which, in the TAC’s judgment, are of adequate quality and scope.
 (from [PQCA project classification](https://docs.google.com/document/d/1NV-0vNgXWdc81oqT0jv0C-9Funb8dySS06u90ghF-X4/edit))

> WE DO NOT CURRENTLY RECOMMEND RELYING ON THIS LIBRARY IN A PRODUCTION ENVIRONMENT OR TO PROTECT ANY SENSITIVE DATA
(from [liboqs usage advice](https://github.com/open-quantum-safe/liboqs?tab=readme-ov-file#limitations-and-security)).

and one unresolved issue [#1](https://github.com/open-quantum-safe/tsc/issues/1) to document that labelling `liboqs` "Impact Stage" as being completely unjustifiable; according to the above, `liboqs` isn't even "Growth Stage". Now, if we were to correct this, not a single OQS sub project would be anywhere else than "Incubation stage" -- pretty much leading the whole PQCA classification ad absurdum -- which is why I'd suggest dropping that column outright (also automatically resolving my initial comments regarding the PQCA classification of `oqs-provider` :-).

Here's btw my practical, non-PQCA-process driven suggestion as to how to handle `oqs-demos` now: https://github.com/open-quantum-safe/oqs-demos/pull/275

**Comment by christianpaquin** [comment]
WRT OpenSSH, I discussed with my AWS peers with whom we do interop testing in the NCCoE project, but there doesn't seem to have bandwidth to take maintenance ownership (@brian-jarvis-aws can confirm)

**Comment by brian-jarvis-aws** [comment]
> WRT OpenSSH, I discussed with my AWS peers with whom we do interop testing in the NCCoE project, but there doesn't seem to have bandwidth to take maintenance ownership (@brian-jarvis-aws can confirm)

That’s correct regarding maintenance, however I do still plan to allocate time for one of my team members to bring the project up to date with the latest upstream OpenSSH version for continued interop testing.

**Comment by baentsch** [comment]
One more criterion helping to define to a sub project's maturity level: [Security incident handling](https://github.com/open-quantum-safe/tsc/issues/60).

**Comment by baentsch** [comment]
To re-activate this issue, here's a straw man proposal for the `oqsprovider` sub project: We don't have to drop it, but everyone shall see (and communicate) it exactly (and "only") as 

> A test & experimentation vehicle for the widest possible range of PQC algorithms, algorithm combinations and use cases. It shall never cater for productive use cases.

Most notably, no-one in PQCA and beyond shall assume or represent this sub-project to be anything else than "Research code" or "Incubation stage" code. Top three reasons:

1) It has so many bells and whistles catering for all kinds of configurations that it's very hard to understand and maintain by now and thus absolutely not suitable for "real use". The good thing is, it does not have to be given the much cleaner code landing now in `openssl`.

2) For a whole year within PQCA, no-one [tried to address issue 1](https://github.com/open-quantum-safe/oqs-provider/issues/375) above or took up [the baton of even defining what it would need to provide "reliability"](https://github.com/open-quantum-safe/oqs-provider/issues/483), let alone contribute work towards that end --short of asking to [simply drop the usage warning](https://github.com/PQCA/TAC/issues/44). The good thing is, addressing these issues is not really essential for "test-and-prototyping-only" code.

3) It has always been an algorithm-agnostic wrapper around `liboqs` functionality -- and there's [calls to change that and adapt specific algorithms to specific use cases and draft specifications](https://github.com/open-quantum-safe/oqs-provider/issues/613) which would make the code ever more "weird" or "cumbersome" -- and those are terms that should arguably not be associated with software handling sensitive data "for real" (ie, and eg. running when your grandma is doing her e-banking :).

Anyone disagreeing with this goal and argumentation, please speak up.

**Comment by dstebila** [comment]
In interest of reviving some of this discussion ahead of next week's TSC meeting, here is an update to https://github.com/open-quantum-safe/tsc/issues/2#issuecomment-2103648852 to reflect what I think is the current status of our sub-projects, according to the criteria in the top post of this thread ("Fully supported (at least 2 fully committed maintainers), best effort (maybe a part-time maintainer), inactive (none of us cares).") plus two new tiers: deprecated/archived, and unclassified

- Fully supported: liboqs, oqs-provider
- Best effort: boringssl, liboqs-cpp, liboqs-go, liboqs-python, oqs-demos
- Inactive: liboqs-java, liboqs-rust, openssh, profiling
- Deprecated/archived: liboqs-dotnet, libssh, openssl-111, oqs-engine
- Unclassified (supporting repositories): .github, ci-containers, liboqs-cupqc, tsc, www

This time I haven't tried classify these against the PQCA tiers since that didn't go well last time and we haven't been asked to do it at this point.

Some questions:

- Are these tiers still meaningful for (a) us in our decision making, and (b) in terms of communicating to the outside world?
- Is the classification of our sub-projects against these tiers accurate?
- Do we want (and are there willing contributors) to move a sub-project from one tier to another?

**Comment by SWilson4** [comment]
>     * Fully supported: liboqs, oqs-provider
> 
>     * Best effort: boringssl, liboqs-cpp, liboqs-go, liboqs-python, oqs-demos
> 
>     * Inactive: liboqs-java, liboqs-rust, openssh, profiling
> 
>     * Deprecated/archived: liboqs-dotnet, libssh, openssl-111, oqs-engine
> 
>     * Unclassified (supporting repositories): .github, ci-containers, liboqs-cupqc, tsc, www

I think this is pretty much accurate, though it seems to me that liboqs-java and liboqs-rust are a notch above openssh and profiling (but a notch below boringssl and the other language wrappers), as both are up to date with liboqs main and actively running in CI. I'm not sure that this warrants moving them up a tier.

> Some questions:
> 
>     * Are these tiers still meaningful for (a) us in our decision making, and (b) in terms of communicating to the outside world?
> 
>     * Is the classification of our sub-projects against these tiers accurate?
> 
>     * Do we want (and are there willing contributors) to move a sub-project from one tier to another?

I would love to see profiling move up a tier. I'm wondering if we should make a pitch to PQCA for some dev cycles from an external contributor to revive and/or replace it, as I'm not sure the core team has the bandwidth to do so at the moment. Maybe something for me to bring up at a TAC call?

**Comment by baentsch** [comment]
> I'm not sure that this warrants moving them up a tier.

I am --> There actually should be an additional tier (2): "Somewhat looked after" (by at least one person) below "Fully supported". This IMO applies to boringssl, liboqs-python (one up each), and oqsprovider (one down).

I'd also agree with moving liboqs-rust and liboqs-java up at least one notch, possibly even to the new tier 2.

Also, one more criterion: "regular releases done" and "inclusion in security incident handling as defined at TSC level". Suggest that as a prerequisite for tier 1 (and possibly the new tier 2 proposed above).

**Comment by dstebila** [comment]
> > I'm not sure that this warrants moving them up a tier.
> 
> I am --> There actually should be an additional tier (2): "Somewhat looked after" (by at least one person) below "Fully supported". This IMO applies to boringssl, liboqs-python (one up each), and oqsprovider (one down).
> 
> I'd also agree with moving liboqs-rust and liboqs-java up at least one notch, possibly even to the new tier 2.
> 
> Also, one more criterion: "regular releases done" and "inclusion in security incident handling as defined at TSC level". Suggest that as a prerequisite for tier 1 (and possibly the new tier 2 proposed above).

So it sounds like the tiers are a little more like: 

- Community supported (maybe "community" supported better represents our current state than "fully" supported?): at least 2 maintainers, plus an active community of contributors; regular releases done; inclusion in security incident handling
- Single maintainer supported: 1 maintainer, with some additional contributors; regular releases done; inclusion in security incident handling
- Best effort (probably could a better name here): Occasional contributors and occasional releases (wording help welcome!)
- Inactive
- Deprecated/archived
- Unclassified supporting repositories
